File: practice/exercism/ruby/linked-list/linked_list.rb
class Deque
  def initialize
    @head = nil
  end

  def node_step_push(head, data)
    current_node = head
    until current_node.next_node == nil
      current_node = current_node.next_node
    end
    current_node.next_node = Node.new(data, current_node)
  end

  def node_step_pop(head)
    current_node = head
    until current_node.next_node.next_node == nil
      current_node = current_node.next_node
    end
    data = current_node.next_node.data
    current_node.next_node = nil
    data
  end

  def push(data)
    if @head.nil?
      @head = Node.new(data)
    else
      node_step_push(@head, data)
    end
  end

  def pop
    if @head.next_node.nil?
      data = @head.data
      @head = nil
    else
      data = node_step_pop(@head)
    end
    data
  end

  def shift
    data = @head.data
    if @head.next_node.nil?
      @head = nil
    else
      @head = @head.next_node
      @head.previous_node = nil
    end
    data
  end

  def unshift(data)
    if @head.nil?
      @head = Node.new(data)
      @head.data
    else
      @head.previous_node = Node.new(data, nil, @head)
      @head = @head.previous_node
    end
  end
end

class Node
  attr_accessor :data, :next_node, :previous_node
  def initialize(data, previous_node = nil, next_node = nil)
    @data = data
    @next_node = next_node
    @previous_node = previous_node
  end
end

File: practice/exercism/javascript/linked-list/linked-list.js
var LinkedList = function() {
  this.head = null;

  this.push = function(node) {
    newNode = new Node(node, null);
    if(this.head) { var node = this.nodeStep(this.head, newNode) : this.head = node
  }

  this.pop = function() {

  }

  this.nodeStep = function(currentNode, node, action) {
    if(currentNode.nextNode) {
      this.nodeStep(currentNode.nextNode, node)
    } else {
      if(action = "push") {
        currentNode.next = node;
        return node;
      } else if(action = "pop") {

      }

    }
  }
}

var Node = function(data, nextNode) {
  this.data = data;
  this.next = nextNode;
}

module.exports = LinkedList, Node
